### PR TITLE
Improve Service controller's code coverage a little bit

### PR DESCRIPTION
**Pull request by dhilipkumars**
**What this PR does / why we need it**:
Improves the code coverage for Service Controller
Before
```
go test --cover ./pkg/controller/service
ok      k8s.io/kubernetes/pkg/controller/service        0.101s  coverage: 23.4% of statements
```
After
```
go test --cover ./pkg/controller/service/
ok      k8s.io/kubernetes/pkg/controller/service        0.094s  coverage: 62.0% of statements
```

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
More unit testing

**Special notes for your reviewer**:

**Release note**: 

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/service/servicecontroller_test.go [hunk 1]
```diff
@@ -296,3 +298,311 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+		{
+			testName: "if an invalid service name is synced",
+			key:      "invalid/key/string",
+			updateFn: func() {
+				controller, cloud, _ = newController()
+
+			},
+			expectedErr: fmt.Errorf("unexpected key format: %q", "invalid/key/string"),
+		},
+		/* We cannot open this test case as syncService(key) currently runtime.HandleError(err) and suppresses frequently occurring errors
```
Comment (cmluciano) [inline pkg/controller/service/servicecontroller_test.go]: Should this be removed?

File: pkg/controller/service/servicecontroller_test.go [hunk 2]
```diff
@@ -296,3 +298,311 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+
+	for _, tst := range testCases {
+		var printError bool
+		tst.updateFn()
+		obtainedErr := controller.syncService(tst.key)
+		if obtainedErr != tst.expectedErr {
+			printError = true
+			if obtainedErr != nil && tst.expectedErr != nil {
+				if obtainedErr.Error() == tst.expectedErr.Error() {
+					//Both are valid error objects but still their content is same, then not an error
```
Comment (cmluciano) [inline pkg/controller/service/servicecontroller_test.go]: This sentence does not read well to me. Can you rephrase to clarify your intent? The comment might also be unnecessary.

File: pkg/controller/service/servicecontroller_test.go [hunk 3]
```diff
@@ -296,3 +298,311 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+
+	testCases := []struct {
+		testName        string                                                //Name of the testcase
+		updateFn        func()                                                //Update function used to manupulate srv and controller values
+		checkExpectedFn func(srvErr error, retryDuration time.Duration) error //Function to check if the returned value is expected
+	}{
+		{
+			testName: "If an invalid service is deleted",
+			updateFn: func() {
+				//srv = defaultExternalService()
```
Comment (cmluciano) [inline pkg/controller/service/servicecontroller_test.go]: Need to remove?

File: pkg/controller/service/servicecontroller_test.go [hunk 4]
```diff
@@ -296,3 +298,311 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+				controller, cloud, _ = newController()
+
+			},
+			checkExpectedFn: func(srvErr error, retryDuration time.Duration) error {
+
+				expectedError := "Service external-balancer not in cache even though the watcher thought it was. Ignoring the deletion."
+				if srvErr == nil || srvErr.Error() != expectedError {
+					//cannot be nil or Wrong error message
```
Comment (cmluciano) [inline pkg/controller/service/servicecontroller_test.go]: Comment may be uneccesary

File: pkg/controller/service/servicecontroller_test.go [hunk 5]
```diff
@@ -296,3 +298,311 @@ func TestGetNodeConditionPredicate(t *testing.T) {
 }
 
 // TODO(a-robinson): Add tests for update/sync/delete.
```
Comment (cmluciano) [inline pkg/controller/service/servicecontroller_test.go]: Can probably remove this TODO if this is merged

**Review by cmluciano — COMMENTED** [review]
Mostly nits:
- I think a lot of the code is readable enough that the comments may not be necessary [Ex. struct comments](https://github.com/kubernetes/kubernetes/pull/41763/files#diff-d57aa6e65ab02bdc723f13236834a5d8R322).
- Some portions are commented out as either not working or are values that may have been used for testing. These should probably be removed in place of TODOs if someone should take a stab at adding them later.

File: pkg/controller/service/servicecontroller_test.go (hunk 5 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: agreed

File: pkg/controller/service/servicecontroller_test.go (hunk 1 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: will remove it, and replace it with a single sentence TODO

File: pkg/controller/service/servicecontroller_test.go (hunk 3 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: agreed

File: pkg/controller/service/servicecontroller_test.go (hunk 4 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: agreed

**Comment by dhilipkumars** [comment]
@cmluciano and @bprashanth 
Shall i also remove this function? Its not being used anywhere
```
func intSlicesEqual(x, y []int) bool {
	if len(x) != len(y) {
		return false
	}
	if !sort.IntsAreSorted(x) {
		sort.Ints(x)
	}
	if !sort.IntsAreSorted(y) {
		sort.Ints(y)
	}
	for i := range x {
		if x[i] != y[i] {
			return false
		}
	}
	return true
}
```

**Comment by dhilipkumars** [comment]
CC other owners for more feedback:@MrHohn @bowei @thockin @mrubin

**Comment by dhilipkumars** [comment]
PING, Hello authors @bprashanth @bowei @thockin @MrHohn @mrubin any feedback on this PR? Its been a month

**Comment by dhilipkumars** [comment]
@k8s-bot ok to test

**Comment by MrHohn** [comment]
@k8s-bot ok to test

/assign

**Comment by MrHohn** [comment]
@dhilipkumars Terribly sorry about the delay. I will review your PR soon.

File: pkg/controller/service/servicecontroller_test.go [hunk 6]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+		expectedErr error  //syncService() only returns error
+	}{
+		{
+			testName: "if an invalid service name is synced",
+			key:      "invalid/key/string",
+			updateFn: func() {
+				controller, cloud, _ = newController()
+
+			},
+			expectedErr: fmt.Errorf("unexpected key format: %q", "invalid/key/string"),
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: This might be risky, it will be broken whenever client-go package changes the error format. Probably enough to just check that it returns error?

File: pkg/controller/service/servicecontroller_test.go [hunk 7]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+				controller, cloud, _ = newController()
+				srv := controller.cache.getOrCreate("external-balancer")
+				srv.state = defaultExternalService()
+			},
+			expectedErr: nil,
+		},
+	}
+
+	for _, tst := range testCases {
+		var printError bool
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: This looks weird to me, could we figure out a simpler way to log error?

File: pkg/controller/service/servicecontroller_test.go [hunk 8]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+	var controller *ServiceController
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName        string
+		updateFn        func()                                                //Update function used to manupulate srv and controller values
+		checkExpectedFn func(srvErr error, retryDuration time.Duration) error //Function to check if the returned value is expected
+	}{
+		{
+			testName: "If an invalid service is deleted",
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: If a non-exist service is deleted

File: pkg/controller/service/servicecontroller_test.go [hunk 9]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+
+func TestDeleteService(t *testing.T) {
+
+	var controller *ServiceController
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName        string
+		updateFn        func()                                                //Update function used to manupulate srv and controller values
+		checkExpectedFn func(srvErr error, retryDuration time.Duration) error //Function to check if the returned value is expected
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: You are using expectedFn elsewhere, why not make it consistent here?
```
expectedFn func(error, time.Duration) error
```

File: pkg/controller/service/servicecontroller_test.go [hunk 10]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+					return fmt.Errorf("RetryDuration Expected=%v Obtained=%v", minRetryDelay, retryDuration)
+				}
+				return nil
+			},
+		},
+		{
+			testName: "If delete was successful",
+			updateFn: func() {
+
+				controller, cloud, _ = newController()
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: Seems like newController() is needed for all test cases. Why not move it out of the function and pass it in asan argument?

File: pkg/controller/service/servicecontroller_test.go [hunk 11]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+	}
+}
+
+func TestDeleteService(t *testing.T) {
+
+	var controller *ServiceController
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName        string
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: Why not also have a key field here, just like TestUpdateService?

File: pkg/controller/service/servicecontroller_test.go [hunk 12]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+				}
+
+				return nil
+			},
+		},
+	}
+
+	for _, tst := range testCases {
+		tst.updateFn()
+		obtainedErr, retryDuration := controller.processServiceDeletion("external-balancer")
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: Add the key field and pass it in here.

File: pkg/controller/service/servicecontroller_test.go [hunk 13]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+
+// Add unit testing for needsUpdate functions
+
+func TestDoesExternalLoadBalancerNeedsUpdate(t *testing.T) {
+
+	var oldSrv, newSrv *v1.Service
+
+	testCases := []struct {
+		testName       string //Name of the test case
+		updateFn       func() //Function to update the old and new service varuables
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: s/varuables/variables

Or "Function to update the service object"

File: pkg/controller/service/servicecontroller_test.go [hunk 14]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+		expectedResult bool   //needsupdate always returns bool
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: expectedResult -> expectNeedsUpdate, or something that is more obvious

File: pkg/controller/service/servicecontroller_test.go [hunk 15]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+
+	}{
+		{
+			testName: "If the service type is different",
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: I think this test name is not accurate? Changes like ClusterIP -> NodePort will not trigger loadBalancer update. Probably just say change service type from LoadBalancer to ClusterIP?

File: pkg/controller/service/servicecontroller_test.go (hunk 13 above)
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: The way you are using updateFn here is not intuitive. Usually I would expect oldService and newService as fields in the testcase struct. Though I could see making this a function make it simpler for tweaking both old and new services. We could keep it for now, but it will be great if you have better idea :)

File: pkg/controller/service/servicecontroller_test.go [hunk 16]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+		obtainedResult := controller.needsUpdate(oldSrv, newSrv)
+		if obtainedResult != tst.expectedResult {
+			t.Errorf("%v needsUpdate() should have returned %v but returned %v", tst.testName, tst.expectedResult, obtainedResult)
+		}
+	}
+}
+
+func TestServiceCache(t *testing.T) {
+
+	sC := &serviceCache{serviceMap: make(map[string]*cachedService)} //ServiceCache
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: nit: sC -> sc

File: pkg/controller/service/servicecontroller_test.go [hunk 17]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+			expectedFn: func() error {
+				//Now it should have no element
+				if len(sC.serviceMap) != 0 {
+					return fmt.Errorf("Expected=0 Obtained=%d", len(sC.serviceMap))
+				}
+				return nil
+			},
+		},
+		{
+			testName: "SetandGet",
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: nit: SetandGet -> Set and Get

File: pkg/controller/service/servicecontroller_test.go [hunk 18]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+func TestUtilityFunctions_nodeSlicesEqualForLB(t *testing.T) {
+	numNodes := 10
+	nArray := make([]*v1.Node, 10)
+
+	for i := 0; i < numNodes; i++ {
+		nArray[i] = &v1.Node{}
+		nArray[i].Name = fmt.Sprintf("node1")
+	}
+	if !nodeSlicesEqualForLB(nArray, nArray) {
+		t.Errorf("Expected=true Obtained=false")
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: Could we also put the function name (nodeSlicesEqualForLB) in the error log to make it clearer?

File: pkg/controller/service/servicecontroller_test.go (hunk 16 above)
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: All the testcases will be sharing this one cache, is this intended? And doe the order of test cases matter? In that case, I think we should at least put a comment so folks won't break it.

File: pkg/controller/service/servicecontroller_test.go [hunk 19]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+}
+
+func TestServiceCache(t *testing.T) {
+
+	sC := &serviceCache{serviceMap: make(map[string]*cachedService)} //ServiceCache
+
+	testCases := []struct {
+		testName   string
+		runTest    func()
+		expectedFn func() error
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: According to how you use runTest and expectedFn, I think they should have more appropriate names. Probably setCacheFn and checkCacheFn (I'm also bad at naming :()?

File: pkg/controller/service/servicecontroller_test.go [hunk 20]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+				srv, isKey, err := sC.GetByKey("addTest")
+				if srv == nil || isKey == false || err != nil {
+					return fmt.Errorf("Expected(non-nil, true, nil) Obtained(%v,%v,%v)", srv, isKey, err)
+				}
+				return nil
+			},
+		},
+		{
+			testName: "allServices",
+			runTest: func() {
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: Probably set it to nil, and check for nil and do nothing in the test case loop?

**Review by MrHohn — CHANGES_REQUESTED** [review]
Thanks for the PR and improving the our coverage rate :)

**Comment by dhilipkumars** [comment]
Thanks for the review i will update the PR addressing your comments.  Thanks.  I will also try include [this](https://github.com/kubernetes/kubernetes/commit/5c9fe166f98136b15a32a7f281d995d1e8a44b63) commit changes,  this PR gives an additional of 15% coverage.

File: pkg/controller/service/servicecontroller_test.go [hunk 21]
```diff
@@ -61,6 +63,20 @@ func newController() (*ServiceController, *fakecloud.FakeCloud, *fake.Clientset)
 	return controller, cloud, client
 }
 
+func defaultExternalService() *v1.Service {
```
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done

File: pkg/controller/service/servicecontroller_test.go [hunk 22]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
 	}
 }
 
-// TODO(a-robinson): Add tests for update/sync/delete.
+func TestUpdateService(t *testing.T) {
```
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done

File: pkg/controller/service/servicecontroller_test.go [hunk 23]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+
+	var controller *ServiceController
+	var cloud *fakecloud.FakeCloud
+
+	testCases := []struct {
+		testName   string
+		key        string
+		updateFn   func() //Manupulate the structure
+		srv        *v1.Service
```
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done throughout the tests

File: pkg/controller/service/servicecontroller_test.go [hunk 24]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+				}
+				if retryDuration != doNotRetry {
+					return fmt.Errorf("retryDuration Expected=%v Obtained=%v", doNotRetry, retryDuration)
+				}
+				return nil
+			},
+		},
+	}
+
+	for _, tst := range testCases {
```
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done throughout the tests

File: pkg/controller/service/servicecontroller_test.go (hunk 6 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: agreed removed, i was thinking if we have more than once situation where it throws error how do we differentiate it, so far we have only one case that throws error

File: pkg/controller/service/servicecontroller_test.go (hunk 7 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: Simplified now

File: pkg/controller/service/servicecontroller_test.go [hunk 25]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+				}
+			}
+		}
+		if printError {
+			t.Errorf("%v Expected=%v Obtained=%v", tst.testName, tst.expectedErr, obtainedErr)
+		}
+	}
+}
+
+func TestDeleteService(t *testing.T) {
```
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done

File: pkg/controller/service/servicecontroller_test.go (hunk 11 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: I have moved the key as a global variable as we are using the same key throughout the test cases.

File: pkg/controller/service/servicecontroller_test.go (hunk 9 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done

File: pkg/controller/service/servicecontroller_test.go (hunk 8 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done

File: pkg/controller/service/servicecontroller_test.go (hunk 10 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done

File: pkg/controller/service/servicecontroller_test.go (hunk 12 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: moved the key as global variable so that it can be changed

File: pkg/controller/service/servicecontroller_test.go [hunk 26]
```diff
@@ -295,4 +311,408 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+		tst.updateFn()
+		obtainedErr, retryDuration := controller.processServiceDeletion("external-balancer")
+		if err := tst.checkExpectedFn(obtainedErr, retryDuration); err != nil {
+			t.Errorf("%v processServiceDeletion() %v", tst.testName, err)
+		}
+	}
+
+}
+
+// Add unit testing for needsUpdate functions
```
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done

File: pkg/controller/service/servicecontroller_test.go (hunk 13 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done

File: pkg/controller/service/servicecontroller_test.go (hunk 14 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done

File: pkg/controller/service/servicecontroller_test.go (hunk 15 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done

File: pkg/controller/service/servicecontroller_test.go (hunk 19 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: _more appropriate names. Probably setCacheFn and checkCacheFn__
Agreed.
 _(I'm also bad at naming :()_
Disagreed.  You are far better. :-)

File: pkg/controller/service/servicecontroller_test.go (hunk 20 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: great idea, done

File: pkg/controller/service/servicecontroller_test.go (hunk 18 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: done, but thought it would come anyways when you run the test with -v?

**Comment by dhilipkumars** [comment]
@MrHohn 
Now after merging [this](https://github.com/kubernetes/kubernetes/commit/5c9fe166f98136b15a32a7f281d995d1e8a44b63) commit we get 3% improved coverage. 
```
time go test --cover ./pkg/controller/service/
ok      k8s.io/kubernetes/pkg/controller/service        0.092s  coverage: 65.0% of statements

real    1m41.993s
user    2m53.844s
sys     0m18.272s

```
PTAL now.

File: pkg/controller/service/servicecontroller_test.go [hunk 27]
```diff
@@ -650,95 +628,100 @@ func TestDoesExternalLoadBalancerNeedsUpdate(t *testing.T) {
...
+		obtainedResult := controller.needsUpdate(oldSvc, newSvc)
+		if obtainedResult != tc.expectedNeedsUpdate {
+			t.Errorf("%v needsUpdate() should have returned %v but returned %v", tc.testName, tc.expectedNeedsUpdate, obtainedResult)
 		}
 	}
 }
 
+//All the testcases for ServiceCache uses a single cache, these below test cases should be run in order,
+//as tc1 (addCache would add elements to the cache)
+//and tc2 (delCache woudl remove element from the cache without it adding automatically)
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: s/woudl/would

File: pkg/controller/service/servicecontroller_test.go [hunk 28]
```diff
@@ -314,139 +307,112 @@ func TestGetNodeConditionPredicate(t *testing.T) {
...
+					return err
+				}
+				if retryDuration != doNotRetry {
+					return fmt.Errorf("retryDuration Expected=%v Obtained=%v", doNotRetry, retryDuration)
+				}
+				return nil
+			},
+		},
+		{
+			testName: "If Updating a  new Loadbalancer IP",
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: tiny nit: "If Updating a  new Loadbalancer IP" -> "If updating LoadBalancer IP"

File: pkg/controller/service/servicecontroller_test.go [hunk 29]
```diff
@@ -484,54 +450,58 @@ func TestSyncService_new(t *testing.T) {
...
-					printError = false
-				}
-			}
+	for _, tc := range testCases {
+
+		tc.updateFn()
+		obtainedErr := controller.syncService(tc.key)
+
+		//Lets only check if we expect error but recived nil and viceversa.
+		if (obtainedErr != nil && tc.expectedErr == nil) && (obtainedErr == nil && tc.expectedErr != nil) {
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: Should use ||?
```
if (obtainedErr != nil && tc.expectedErr == nil) || (obtainedErr == nil && tc.expectedErr != nil)
```

File: pkg/controller/service/servicecontroller_test.go (hunk 6 above)
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: Fair point, though I'm still concern about the consistence of error message, which depends on other package. Or could we make this a boolean for now and put a TODO?

File: pkg/controller/service/servicecontroller_test.go [hunk 30]
```diff
@@ -484,54 +450,58 @@ func TestSyncService_new(t *testing.T) {
...
+		_, exist := controller.cache.get(tc.key)
+		if exist {
+			t.Fatalf("sync service error, workingQueue should not contain service: %s any more", tc.key)
 		}
 	}
+
 }
 
-func TestDeleteService(t *testing.T) {
+func TestProcessDeleteService(t *testing.T) {
```
Comment (MrHohn) [inline pkg/controller/service/servicecontroller_test.go]: Make it TestProcessServiceDeletion to be consistent with the origin function name.

**Review by MrHohn — COMMENTED** [review]
Thanks! Mostly looks good, put some nits and comments.

**Comment by MrHohn** [comment]
BTW, could you change the release note to:

**Release note**:

```release-note
NONE
```

So the robot will replace the release note label.

File: pkg/controller/service/servicecontroller_test.go (hunk 28 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: Agreed

File: pkg/controller/service/servicecontroller_test.go (hunk 29 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: Oops, that is correct.

File: pkg/controller/service/servicecontroller_test.go (hunk 6 above)
Comment (dhilipkumars) [inline pkg/controller/service/servicecontroller_test.go]: Done, Added TODO and chacked it to only check for error.  This helps a little more for us to print what was expected and what was obtained.

**Comment by dhilipkumars** [comment]
@k8s-bot 
Release Notes: None

**Comment by dhilipkumars** [comment]
```
BTW, could you change the release note to:
```
Should i be part of the Org to be able to do that? Im unable to change or add any lables.

**Comment by MrHohn** [comment]
> Should i be part of the Org to be able to do that? Im unable to change or add any lables.

No, you don't need to in this case. Once you change the release note in your first post to NONE, bot will apply label for you. As I can see now the release note is still empty, instead of NONE.

**Review by MrHohn — APPROVED** [review]
Thanks! LGTM. Could you squash the commits?

**Comment by dhilipkumars** [comment]
done now

**Comment by MrHohn** [comment]
/assign @thockin
for approval.
